src/databricks/labs/ucx/source_code/pyspark.py

```diff
@@ -1,10 +1,155 @@
 import ast
-from collections.abc import Iterable
+from abc import ABC, abstractmethod
+from collections.abc import Iterable, Iterator
+from dataclasses import dataclass
 
-from databricks.labs.ucx.source_code.base import Advice, Fixer, Linter
+from databricks.labs.ucx.source_code.base import Advice, Deprecation, Fixer, Linter
 from databricks.labs.ucx.source_code.queries import FromTable
 
 
+@dataclass
+class Matcher(ABC):
+    method_name: str
+    min_args: int
+    max_args: int
+    table_arg_index: int
+
+    def matches(self, node: ast.AST):
+        return (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Attribute)
+            and self.min_args <= len(node.args) <= self.max_args
+            and isinstance(node.args[self.table_arg_index], ast.Constant)
+        )
+
+    @abstractmethod
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def apply(self, from_table: FromTable, node: ast.Call) -> None:
+        raise NotImplementedError()
+
+
+@dataclass
+class QueryMatcher(Matcher):
+
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        table_arg = node.args[self.table_arg_index]
```

Comment:
what if the argument is passed by name? do we want to defensively add an argument name to a matcher?

Reply:
you mean by keyword ? like spark.sql.table(tablename="old.thing") ? At this point I believe the `tablename` arg is always positional in the Spark API, so it wouldn't affect the usage of table_arg_index but I'll double check and add a test.

Reply:
tracking in https://github.com/databrickslabs/ucx/issues/1220 to possibly be addressed later. Junior data scientists may abuse kvargs for no good reason and we need to defend against that.

Reply:
Done